broker.go

```diff
@@ -184,25 +306,87 @@ func (b *Broker) RegisterPipeline(def Pipeline) error {
 		return err
 	}
 
+	// Store the pipeline and then update the reference count of the nodes in that pipeline.
 	g.roots.Store(def.PipelineID, root)
+	for _, id := range def.NodeIDs {
+		nodeUsage, ok := b.nodes[id]
+		// We can be optimistic about this as we would have already errored above.
+		if ok {
+			nodeUsage.referenceCount++
+		}
+	}
 
 	return nil
 }
 
 // RemovePipeline removes a pipeline from the broker.
 func (b *Broker) RemovePipeline(t EventType, id PipelineID) error {
+	if t == "" {
+		return errors.New("event type cannot be empty")
+	}
+	if id == "" {
+		return errors.New("pipeline ID cannot be empty")
+	}
+
 	b.lock.Lock()
 	defer b.lock.Unlock()
 
 	g, ok := b.graphs[t]
 	if !ok {
-		return fmt.Errorf("No graph for EventType %s", t)
+		return fmt.Errorf("no graph for EventType %s", t)
 	}
 
 	g.roots.Delete(id)
 	return nil
 }
 
+// RemovePipelineAndNodes will attempt to remove all nodes referenced by the pipeline.
+// Any nodes that are referenced by other pipelines will not be removed.
+func (b *Broker) RemovePipelineAndNodes(t EventType, id PipelineID) error {
```

Comment:
do we want to test the validation errors?

Reply:
`TestRemovePipelineAndNodes_BadParameters` should cover this.